File: bin/dsadmin.js
#!/usr/bin/env node
/* eslint-env node */
/* eslint-disable @typescript-eslint/no-var-requires */

const static = require("node-static");
const process = require("process");
const http = require("http");
const path = require("path");
const proxy = require("http2-proxy");
const fs = require("fs");
const jsesc = require("jsesc");
const yargs = require("yargs/yargs");
const { exit } = require("yargs");

const publicDir = path.join(__dirname, "..", "public");

const args = yargs(process.argv)
  .option("project", {
    type: "string",
    default: process.env.DATASTORE_PROJECT_ID,
    description:
      "Datastore project ID. Defaults to DATASTORE_PROJECT_ID environment variable.",
    demandOption: true,
  })
  .option("datastore-emulator-host", {
    type: "string",
    default: process.env.DATASTORE_EMULATOR_HOST || "127.0.0.1:8081",
    description:
      "Datastore emulator hostname & port. Defaults to DATASTORE_EMULATOR_HOST environment variable.",
    demandOption: true,
  })
  .option("base-path", {
    type: "string",
    default: process.env.DATASTORE_BASEPATH || "",
    description:
      "Base HTTP path to serve the interface on. Defaults to DATASTORE_BASEPATH environment variable.",
  })
  .option("port", {
    type: "number",
    default: 8080,
    description: "Port to listen on",
  }).argv;

const emulatorHost = args.datastoreEmulatorHost.split(":");
if (emulatorHost.length != 2) {
  console.error("Invalid emulator host");
  exit(-1);
}
emulatorHost[1] = parseInt(emulatorHost[1], 10);

const basePath = args.basePath;

const dsadminEnv = {
  DATASTORE_PROJECT_ID: args.project,
  BASE_PATH: basePath,
};

const index = fs
  .readFileSync(path.join(publicDir, "index.html"))
  .toString()
  .replace(
    "<body>",
    `<body><script>DSADMIN_ENV = ${jsesc(dsadminEnv, {
      isScriptContext: true,
    })}</script>`,
  )
  .replaceAll("{{BASE}}", basePath);

function serveIndex(req, res) {
  res.writeHead(200);
  res.end(index);
}

var file = new static.Server(publicDir);

console.log(
  `dsadmin (project ${args.project}) listening on http://localhost:${args.port}${basePath}`,
);
http
  .createServer(function (req, res) {
    if (!req.url.startsWith(basePath)) {
      res.writeHead(400);
      res.end("not found");
      return;
    }
    req.url = req.url.substring(basePath.length);
    if (req.url.startsWith("/v1/")) {
      proxy.web(req, res, {
        hostname: emulatorHost[0],
        port: emulatorHost[1],
      });
    } else if (req.url === "/" || req.url === "/index.html") {
      serveIndex(req, res);
    } else {
      file.serve(req, res, function (e) {
        if (e && e.status === 404) {
          serveIndex(req, res);
        }
      });
    }
  })
  .listen(args.port);

File: cmd/dsadmin/dsadmin.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"html/template"
	"io"
	"io/fs"
	"log"
	"net/http"
	"net/http/httputil"
	"os"

	"github.com/remko/dsadmin"
)

func NewAssetFS(root fs.FS, fallback []byte) http.Handler {
	assets := http.FileServer(http.FS(root))
	return http.StripPrefix("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if _, err := fs.Stat(root, r.URL.Path); err != nil {
			w.Write(fallback)
			return
		}
		assets.ServeHTTP(w, r)
	}))
}

func main() {
	project := flag.String("project", os.Getenv("DATASTORE_PROJECT_ID"), "Datastore project ID. Defaults to DATASTORE_PROJECT_ID environment variable.")
	port := flag.Uint("port", 8080, "Port to listen on")
	basePath := flag.String("base-path", "", "Base HTTP path to serve the interface on")
	defaultEmulatorHost := os.Getenv("DATASTORE_EMULATOR_HOST")
	if defaultEmulatorHost == "" {
		defaultEmulatorHost = "localhost:8081"
	}
	emulatorHost := flag.String("datastore-emulator-host", defaultEmulatorHost, "Datastore emulator hostname & port. Defaults to DATASTORE_EMULATOR_HOST environment variable.")
	flag.Parse()
	if *project == "" {
		log.Fatalf("missing project setting")
	}
	if *emulatorHost == "" {
		log.Fatalf(("missing datastore emulator host setting"))
	}

	f, err := dsadmin.WebAppAssets.Open("index.html")
	if err != nil {
		panic(err)
	}
	index, err := io.ReadAll(f)
	if err != nil {
		panic(err)
	}
	dsadminEnv, err := json.Marshal(map[string]interface{}{
		"DATASTORE_PROJECT_ID": project,
		"BASE_PATH":            basePath,
	})
	if err != nil {
		panic(err)
	}
	index = bytes.Replace(index, []byte("<body>"), []byte(fmt.Sprintf("<body><script>DSADMIN_ENV = JSON.parse(\"%s\");</script>", template.JSEscapeString(string(dsadminEnv)))), 1)
	index = bytes.Replace(index, []byte("{{BASE}}"), []byte(*basePath), -1)

	http.Handle(*basePath+"/v1/", http.StripPrefix(*basePath, &httputil.ReverseProxy{
		Director: func(req *http.Request) {
			req.URL.Scheme = "http"
			req.URL.Host = *emulatorHost
			req.Host = *emulatorHost
		}},
	))
	http.Handle(*basePath+"/", http.StripPrefix(*basePath, NewAssetFS(dsadmin.WebAppAssets, index)))

	log.Printf("dsadmin (project '%s') listening on http://localhost:%d%s", *project, *port, *basePath)
	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%d", *port), nil))
}
